File: app/src/main/java/com/mstc/mstcapp/ui/ProjectIdeaFragment.java
package com.mstc.mstcapp.ui;

import android.app.ProgressDialog;
import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.mstc.mstcapp.R;
import com.mstc.mstcapp.model.ProjectIdeaModel;
import com.mstc.mstcapp.util.RetrofitInstance;
import com.mstc.mstcapp.util.RetrofitInterface;

import java.util.Map;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class ProjectIdeaFragment extends BottomSheetDialogFragment {
    private static final String TAG = "ProjectIdeaFragment";
    private Context context;
    private TextInputEditText name, email, phone, idea, description;
    private TextInputLayout name1, email1, phone1, idea1, description1;
    private ProgressDialog progressDialog;

    public ProjectIdeaFragment() {
    }

    public static ProjectIdeaFragment newInstance() {
        return new ProjectIdeaFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        context = getContext();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_project_idea, container, false);
        name = view.findViewById(R.id.name);
        email = view.findViewById(R.id.email);
        phone = view.findViewById(R.id.phone);
        idea = view.findViewById(R.id.idea);
        description = view.findViewById(R.id.description);
        name1 = view.findViewById(R.id.name1);
        email1 = view.findViewById(R.id.email1);
        phone1 = view.findViewById(R.id.phone1);
        idea1 = view.findViewById(R.id.idea1);
        description1 = view.findViewById(R.id.description1);
        progressDialog = new ProgressDialog(context);
        progressDialog.setMessage("Posting Idea...");
        name.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (count == 0) name1.setError("Cannot be empty!");
                else name1.setError(null);
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });
        phone.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (count == 0) phone1.setError("Invalid Phone Number!");
                else phone1.setError(null);
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (s.length() != 10) phone1.setError("Invalid Phone Number!");
                else phone1.setError(null);
            }
        });

        email.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (!s.toString().matches("^[a-zA-Z0-9+_.-]+@[a-zA-Z0-9.-]+$"))
                    email1.setError("Invalid Email Address");
                else email1.setError(null);
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });


        view.findViewById(R.id.post).setOnClickListener(v -> {
            ProjectIdeaModel projectIdeaModel = new ProjectIdeaModel(
                    Objects.requireNonNull(name.getText()).toString(),
                    Objects.requireNonNull(phone.getText()).toString(),
                    Objects.requireNonNull(email.getText()).toString(),
                    Objects.requireNonNull(idea.getText()).toString(),
                    Objects.requireNonNull(description.getText()).toString());
            if (projectIdeaModel.getName().length() == 0)
                name1.setError("Cannot be empty!");
            else if (projectIdeaModel.getPhone().length() == 0)
                phone1.setError("Cannot be empty!");
            else if (projectIdeaModel.getPhone().length() != 10)
                phone1.setError("Invalid Phone Number");
            else if (projectIdeaModel.getEmail().length() == 0)
                email1.setError("Cannot be empty!");
            else if (!projectIdeaModel.getEmail().matches("^[a-zA-Z0-9+_.-]+@[a-zA-Z0-9.-]+$"))
                email1.setError("Invalid Email Address");
            else if (projectIdeaModel.getIdea().length() == 0)
                idea1.setError("Cannot be empty!");
            else if (projectIdeaModel.getDescription().length() == 0)
                description1.setError("Cannot be empty!");
            else
                postData(projectIdeaModel);
        });
        return view;
    }

    private void postData(ProjectIdeaModel projectIdeaModel) {
        progressDialog.show();
        Retrofit retrofit = RetrofitInstance.getRetrofitInstance();
        RetrofitInterface retrofitInterface = retrofit.create(RetrofitInterface.class);
        Call<Map<String, String>> call = retrofitInterface.postIdea(projectIdeaModel);
        call.enqueue(new Callback<Map<String, String>>() {
            @Override
            public void onResponse(@NonNull Call<Map<String, String>> call,
                                   @NonNull Response<Map<String, String>> response) {
                progressDialog.dismiss();
                Log.d(TAG, "onResponse() returned: " + response.body());
                Map<String, String> map = response.body();
                if (map != null)
                    if (Objects.equals(map.get("statusCode"), "201")) {
                        dismiss();
                        new MaterialAlertDialogBuilder(context)
                                .setTitle("Idea Posted Successfully")
                                .setMessage("Your idea has been posted successfully! " +
                                        "Please feel free to check out the resources while we contact you.")
                                .setPositiveButton("Dismiss", (dialog, which) -> dialog.dismiss())
                                .show();
                    } else
                        Toast.makeText(context, "Could not post idea! Try Again", Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onFailure(@NonNull Call<Map<String, String>> call, @NonNull Throwable t) {
                progressDialog.dismiss();
                Log.e(TAG, "onFailure: ", t);
            }
        });
    }
}

File: app/src/main/java/com/mstc/mstcapp/ui/ProjectIdeaFragment.kt
package com.mstc.mstcapp.ui

import android.content.DialogInterface
import android.content.res.ColorStateList
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.mstc.mstcapp.R
import com.mstc.mstcapp.databinding.DialogProjectIdeaBinding
import com.mstc.mstcapp.model.ProjectIdea
import com.mstc.mstcapp.util.RetrofitService
import kotlinx.coroutines.launch
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit


private const val TAG = "ProjectIdeaFragment"

class ProjectIdeaFragment : BottomSheetDialogFragment() {
    lateinit var binding: DialogProjectIdeaBinding
    private val scheduler: ScheduledExecutorService by lazy { Executors.newSingleThreadScheduledExecutor() }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        binding = DialogProjectIdeaBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.bindState()
    }

    private fun DialogProjectIdeaBinding.bindState() {
        post.setOnClickListener {
            ProjectIdea(
                etName.text.toString(),
                etPhone.text.toString(),
                etEmail.text.toString(),
                etIdea.text.toString(),
                etDescription.text.toString()
            ).apply {
                when {
                    name.isEmpty() -> etName.error = "Cannot be empty!"
                    phone.isEmpty() -> etPhone.error = "Cannot be empty!"
                    phone.length != 10 -> etPhone.error = "Invalid Phone Number"
                    email.isEmpty() -> etEmail.error = "Cannot be empty!"
                    !email.matches(
                        Regex("^[a-zA-Z0-9+_.-]+@[a-zA-Z0-9.-]+$")
                    ) -> etEmail.error = "Invalid Email Address"
                    idea.isEmpty() -> etIdea.error = "Cannot be empty!"
                    description.isEmpty() -> etDescription.error =
                        "Cannot be empty!"
                    else -> lifecycleScope.launch { postData(this@apply) }
                }
            }
        }
    }

    private suspend fun postData(projectIdeaModel: ProjectIdea) {
        animateUpload()
        try {
            val service: RetrofitService = RetrofitService.create()
            val response = service.postIdea(projectIdeaModel)
            if (response.isSuccessful) {
                Log.d(TAG, "postData() returned: ${response.body()}")
                dismiss()
                MaterialAlertDialogBuilder(requireContext())
                    .setTitle("Idea Posted Successfully")
                    .setMessage(
                        "Your idea has been posted successfully! " +
                                "Please feel free to check out the resources while we contact you."
                    )
                    .setPositiveButton("Dismiss") { dialog: DialogInterface, _: Int -> dialog.dismiss() }
                    .show()
            } else {
                Toast.makeText(context, "Could not post idea! Try Again", Toast.LENGTH_SHORT).show()
                Log.e(TAG, "postData: Failed! ${response.code()}")
                throw Error(response.message())
            }
        } catch (e: Exception) {
            Log.e(TAG, "postData: ", e)
            Toast.makeText(context, "Could not post idea! Try Again", Toast.LENGTH_SHORT).show()
            binding.apply {
                post.text = getString(R.string.post)
                post.isEnabled = true
                post.strokeColor =
                    ColorStateList.valueOf(
                        ContextCompat.getColor(
                            requireContext(),
                            R.color.colorPrimary
                        )
                    )
                scheduler.shutdown()
            }
        }
    }

    private fun animateUpload() {
        binding.apply {
            post.isEnabled = false
            post.strokeColor =
                ColorStateList.valueOf(ContextCompat.getColor(requireContext(), R.color.gray))
            var count = 0
            try {
                scheduler.scheduleWithFixedDelay({
                    post.text = "POSTING${".".repeat((++count % 3) + 1)}"
                }, 0, 500, TimeUnit.MILLISECONDS)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}